framework/include/image_resource.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+#include <gvk.hpp>
+
+namespace gvk
+{
+	class image_resource_t;
+	class image_resource_impl_t;
+
+	typedef std::unique_ptr<image_resource_t> image_resource;
```

Comment:
Regarding the `std::unique_ptr`, see the comment above. It should actually be a `gvk::owning_resource<T>` but I will have to look into that again.

I haven't found a definition of resources following the link but think that `image_resource_t` does qualify as a resource, since it contains and manages actual data (via a pointer-to-implementation), and there should never be more than one instance for the same resource (data from texture file). I don't think that "config" or "meta" apply here, as the class does not change anything about the resulting image or cubemap. It simply loads the necessary data from one of several backends (image libraries).

There were two considerations that led to this implementation, following the requirement that existing code continues to work as before:

1. The image library that is used to load the texture is neither known at compile time nor when e.g. `create_cubemap_from_file()` is called. This way a user does not have to worry about which library to use. When the texture data is needed, each library tries to open the file in a Chain of Responsibility pattern. This leads to the Bridge Pattern: decoupling the implementor (loading the texture with a library) from the abstraction (access to texture data). The abstract interface for both implementor and abstraction is `image_resource_base_t`, the abstraction that users see is `image_resource_t` (`image_resource` should be the `owning_resource<T>` wrapper), and the actual implementors (for GLI and stbi) have `image_resource_base_t` as their base class. I'd also suggest to rename these classes but haven't decided yet what the most expressive names would be.
2. Because of caching, it isn't known if the texture file should be loaded at all, or if its contents are read from the cache. For this reason, it is not possible to determine which library to load the texture with when an `image_resource` is created. This has to be deferred to a separate `load()` function, which is only called if caching is not used.

The separation into abstraction and implementors meets both requirements and makes the code much easier to implement, read, and maintain. For example, to add a new image library it is only necessary to write a new class with a small set of functions. No changes to `create{image,cubemap}_from_file()` are needed.

Reply:
Whenever I refer to "resource" I actually mean a "Vulkan GPU resource" => e.g. a `avk::image` which is created with `avk::root::create_image` _on the GPU_, i.e. the resource is created on the GPU during that method invocation; and it is destroyed in `avk::image_t::~image_t`. This is the usual _Auto-Vk_ "resource" pattern (always referring to a resource on the GPU, interfaced through the Vulkan API).

I don't think that `image_resource_t` qualifies as such a resource because it only contains actually the meta data of an image (or multiple image files) which are to be loaded or are already loaded... from a Vulkan GPU resource's point of view. But maybe "meta data" is not totally appropriate here as well. `image_resource_t` is rather a "CPU resource" but it is not a "Vulkan GPU resource". This is what I meant when I wrote that the type should be renamed to something other than "resource" to avoid confusion with "Vulkan GPU resources".

So actually, an `image_resource_t` is something temporary, right? I.e. it must only be created during initialization and can be disposed of as soon as the `avk::image` has been created, right? Under this light, I think the usage of a `std::unique_ptr` would still be appropriate in this case since the requirements are different to "Vulkan GPU resources" which must use `avk::owning_resource` for consistency and performance reasons. 

The performance reasons are that with `avk::owning_resource`, a resource can live entirely on the stack. E.g. 
```
std::array<avk::image, 8> eightImagesOnTheStack;
```
does not allocate anything on the heap. A `std::unique_ptr` always performs a heap allocation.

The abstraction layer implemented for `image_resource_t` is fine. I just think that the interface could be improved in terms of usability - and optimally according to a "config" or "meta" pattern... or if not exactly that then something that feels similar in terms of usage.

Reply:
The `image_resource_t` represents image data (usually loaded from a file, but e.g. on Android loading from an .apk would be possible too) that is then transferred to an `avk::image`. It can be discarded once the image is created. Usually it does not have to be created explicitly, but can be created e.g. by `create_cubemap_from_file()` as needed (I have also changed the example to call this function instead of `create_cubemap_from_image_resource()`).

In that regard it is not a resource in the framework, as it is not managed by Vulkan or the GPU. I would suggest to rename it to `image_data_t`, to indicate that it represents the data stored in an image, and rename related types accordingly. 

There should never be more than one instance for each image file, so move-only semantics plus references also work well for this type. The `std::unique_ptr` is actually not needed, this was just a left-over from my first draft, so I think the `avk::owning_resource` is a good choice, and does not limit the usability.

Reply:
Issue #50 deals with refactoring the interface to `create_image_from_file`. Once the cubemap implementation is finished I could work on that, and also a new image loader (#59), which might affect the interface.